### PR TITLE
rand: don't block before random pool is initialized

### LINKED ISSUE #32953: Default hashmap implementation can cause processes to block
When the /dev/urandom pool has not yet been initialized, the use of the default HashMap can cause a process to block.  The behavior is problematic when a program is started by an initscript and is not intended to run in the background.  The process will block until the urandom pool is initialized which on some embedded systems may be several minutes.  

A concrete example of the issue in BurntSushi/docopt.rs#178, but the issue can also occur in other places like regex.

### PR DESCRIPTION
**Pull request by cardoe**
If we attempt a read with getrandom() on Linux the syscall can block
before the random pool is initialized unless the GRND_NONBLOCK flag is
passed. This flag causes getrandom() to instead return EAGAIN while the
pool is uninitialized. To avoid downstream users of crate or std
functionality that have no ability to avoid this blocking behavior this
change causes Rust to read bytes from /dev/urandom while getrandom()
would block and once getrandom() is available to use that. Fixes #32953.

Signed-off-by: Doug Goldstein cardoe@cardoe.com

### CODE REVIEW

**Comment by cardoe** [comment]
More of a RFC than a real PR.

File: src/libstd/sys/unix/rand.rs [hunk 1]
```diff
@@ -63,6 +65,11 @@ mod imp {
                 let err = errno() as libc::c_int;
                 if err == libc::EINTR {
                     continue;
+                } else if err == libc::EAGAIN {
+                    let reader = File::open("/dev/urandom")?;
```
Comment (sfackler) [inline src/libstd/sys/unix/rand.rs]: getrandom_fill_bytes doesn't return a Result, so `?` won't work here.

**Comment by rkruppe** [comment]
Presumably this is "more of an RFC" because it could be quite controversial. If so, could you lay out the motivation and drawbacks, RFC-style (if shorter and more informally)?

I suspect the main drawback is that this may draw from a PRNG that has not yet collected sufficient entropy, which would be bad for predictability reasons. As for motivation, I've not heard complaints about `getentropy` blocking, since AFAIK it only happens early after booting up and never after `getrandom` already succeeded once. Is that correct, do you have anything to add?

**Comment by cardoe** [comment]
@rkruppe If you follow #32953 in turn docopt/docopt.rs#178 you'll see the reason is that for early boot `getrandom()` will block until the pool is initialized with the current Rust implementation. Rust's HashMap seeds its hasher with `getrandom()` since there is no knowledge if the user will be using a bounded data set or an unbounded data set. An unbounded data set would make the HashMap vulnerable to a denial of service while a bounded data set would not. The referenced tickets proposed a few solutions such as allowing crates to opt into a different hasher or a not guaranteed random seed or make the default Rust implementation a best effort. The discussion gravitated towards best effort so I tried to put together a suggestion implementing best effort.

I will also point out that the delay for an initialized random pool on some hardware and some platforms is not a non-trivial amount of time (order of minutes) so it precludes a number of Rust crates (docopt and regex are two I know about currently) from being used in Rust applications that will be used in early boot.

As far as `getentropy()` goes I'm not able to determine from a quick search if it blocks by default or not. But I'm willing to bet that most people are not attempting to write programs used in early boot or their init system in Rust so they are not running into this issue.

The argument against this change was that many users of HashMap do not have unbounded data sets and as a result are paying a penalty by using SipHash over FNV since FNV is more performant than SipHash (ignoring the random initializing penalty) and perhaps people should be educated about how to use HashMap for their use case and select the right hasher. The issue there is that people may not know how their crate will be used and could unintentionally expose their crate's users to a vulnerability. This argument appeared to be rejected in the favor of best effort which matches how systemd works.

**Comment by alexcrichton** [comment]
Thanks for the PR @cardoe! In terms of implementation I'd probably say that because `OsRng::new` returns `io::Result` we shouldn't have the `.unwrap()` for opening `/dev/urandom` in theory, but rather all Linux `OsRng` instances have an open handle to `/dev/urandom` (as they may need it) and then just attempt to use `getrandom` (nonblocking) first.

I'm gonna tag this as `T-libs` to ensure it comes up during triage, I'm sure there'll be some opinons!

cc @rust-lang/libs

File: src/libstd/sys/unix/rand.rs [hunk 2]
```diff
@@ -63,6 +65,11 @@ mod imp {
                 let err = errno() as libc::c_int;
                 if err == libc::EINTR {
                     continue;
+                } else if err == libc::EAGAIN {
+                    let reader = File::open("/dev/urandom").expect("Unable to open /dev/urandom");
+                    let mut reader_rng = ReaderRng::new(reader);
+                    reader_rng.fill_bytes(& mut v[read..]);
+                    read += v.len() as usize;
```
Comment (bluss) [inline src/libstd/sys/unix/rand.rs]: Needs small comment with an explanation.

**Comment by brson** [comment]
lgtm. I feel like the HashMap docs should say something about this, but I also don't want to make a big deal out of it. Nobody should not use HashMap because of this.

Is this also happening in the public rand crate?

**Comment by tbu-** [comment]
"Use bad random numbers if we can't acquire good ones."

**Comment by tbu-** [comment]
@alexcrichton That would block the access to random numbers on systems with `getrandom` if the file descriptor limit is exhausted.

**Comment by tbu-** [comment]
From http://www.2uo.de/myths-about-urandom/:
> /dev/urandom isn't perfect. The problems are twofold:

>On Linux, unlike FreeBSD, /dev/urandom never blocks. Remember that the whole security rested on some starting randomness, a seed?

>Linux's /dev/urandom happily gives you not-so-random numbers before the kernel even had the chance to gather entropy. When is that? At system start, booting the computer. 

As I understand it, this pull request would mean that we sidestep the protection `getrandom(2)` offers and use "not-so-random" numbers for providing the program with supposedly cryptographically secure random numbers.

File: src/libstd/sys/unix/rand.rs (hunk 2 above)
Comment (cardoe) [inline src/libstd/sys/unix/rand.rs]: Please let me know if 61cbd07 is sufficient and I will squash it in.

**Comment by cardoe** [comment]
@brson I've added 121225f which hopefully adds some good user visible information in HashMap. Let me know if that's good and I can squash it in.

As far as the public rand crate goes, I was waiting to see how this progressed but planned on submitting something to that as well.

**Comment by cardoe** [comment]
@tbu- I'd recommend reading my long explanation that I gave @rkruppe above and followed the two referenced issues. I've rehashed the reasons and the options forward in ways I feel are pretty clear between both tickets and this PR. This PR implements the option that seemed favorable by the Rust maintainers. Obviously if a different option is deemed more favorable I will implement that as well.

**Comment by cardoe** [comment]
@alexcrichton I see two issues with always having /dev/urandom opened is we're consuming a fd from a process always which might not be desirable. And we're back to being dependent on having /dev/urandom be open-able which one of the reasons people tried to get away from using it, specifically the LibreSSL people. See https://lwn.net/Articles/606141/ for some context.

**Comment by bluss** [comment]
@tbu- The only use for this Rng is keys for the HashMap hash function.

**Comment by cardoe** [comment]
@alexcrichton Does that reasoning make sense as to why we don't want to always have `/dev/urandom` opened? @brson Do the two addon commits address your review adequately?

**Comment by sfackler** [comment]
We might want to consider expanding our set of platform specific versions of this logic - we could use `SecRandomCopyBytes` on iOS and OSX for example.

**Comment by tbu-** [comment]
@bluss Fair enough. But it should probably still be pointed out that we use low-quality random numbers for the hash map if there aren't any high-quality random numbers available.

I don't know how that affects the attack surface, but generally Rust had chosen the most secure variant that is still practical (for hashmaps), e.g. choosing a secure hasher over a fast one.

**Comment by cardoe** [comment]
@sfackler The code already does that. Look farther down in the source file.

**Comment by sfackler** [comment]
Oh great!

**Comment by cardoe** [comment]
@tbu- I added documentation to that effect in 121225f, are you not happy with the wording? The attack surface is not really affected because the only time a hash map is vulnerable is if the data set is unbounded. Most of the time applications that allow unbounded data sets (e.g. network daemons) are going to be started up after enough hardware (an interrupt source such as a network card). The result here is that it doesn't matter. Plus a weaker random value would still have to be guessed by someone over the network to be able to attack the hash map. Honestly I would recommend doing some reading about the attack surface of hash maps, what and how they need to be seeded to be secure in those cases, the Linux random number generator implementation and the use cases of how this code will be used because you're honestly spreading a little bit of FUD here.

**Comment by alexcrichton** [comment]
@cardoe 

> Does that reasoning make sense as to why we don't want to always have /dev/urandom opened?

Yes, albeit a seemingly much bigger hammer than is necessary here for just a pair of keys to hash maps.

In any case, though, the PR looks ok to me, just wanna make sure that others in @rust-lang/libs are also signed off.

**Comment by brson** [comment]
@cardoe Yes, thank you.

**Comment by alexcrichton** [comment]
The libs team discussed this PR during triage today and the decision was to merge. We may want to comment the internal rand library to indicate that it's not suitable for anything but what the standard library needs (for implementation details such as this). We also discussed panicking here if opening `/dev/urandom` fails, and the conclusion was that this is probably fine for now and we can revisit later if need be.

Thanks again for the PR @cardoe!

@bors: r+ 61cbd07dec2d9670f170bed44f1815ea29a8a214

**Comment by tbu-** [comment]
@alexcrichton Are you aware of the fact that `OsRng` is directly exposed to users of the standard library via `std::rand::OsRng` and used to provide cryptographically random numbers before this commit? Maybe only the random number generation for `HashMap`s should be changed.

**Comment by huonw** [comment]
@tbu- AFAICT, it is private, and would be unstable anyway; could you clarify what you mean?

**Comment by tbu-** [comment]
@huonw Ah right. Sorry, I messed up here, I didn't see that `mod rand` isn't exported by `Std`. I thought we were exposing random number generators previously and now changed them to yield not-so-random numbers.

**Comment by cardoe** [comment]
@tbu- The point of this change was to keep the scope limited just to the internal part of `std` (e.g. HashMap) and not exported for use. If I changed the `rand` crate then yes it would be exposed to other users. But again the scope of the change is extremely limited, even if we did change the `rand` crate. I really encourage you to read about how the Linux `getrandom()` syscall works, the Linux random number generator and follow the code flow of the fallbacks and see that once the urandom pool is initialized this change has no effect.

**Comment by cardoe** [comment]
@alexcrichton Thank you! If anyone on the libs team wants to see a behavior change here I will gladly do an update.

**Comment by alexcrichton** [comment]
@bors: retry

* #33434

**Comment by eternaleye** [comment]
I realize I'm a bit late on this, but I wanted to explain why I feel this is a poor decision, and one based on a false dichotomy.

The problem as expressed is (to my understanding):
- `HashMap` initialization requires randomness for initializing SipHash, to avoid hash-DOS
- Acquiring randomness may block during early boot when using `getrandom()`
- Thus, using `HashMap` during early boot may cause blocking
- Thus, Rust is ill-prepared for early-boot environments

This is justified by that in the fallback case to `/dev/urandom`, such blocking does not occur, and thus the security benefits of it can't be guaranteed.

However:
- Security is best served not by taking the common subset of behaviors, but by choosing the most secure option available at any given time. If this was not true, `getrandom()` would never have blocked in the first place. As "`getrandom()` does not exist" returns a distinct error from "the entropy pool is not initialized", I feel the justification is specious.
- `HashMap` is not the only map available, and `BTreeMap` does not require randomness to avoid DOS.
- Early-boot environments are special; expecting arbitrary code written for normal environments to be sensible for early-boot unmodified is fallacious. Thus, recommending the use of a different map is reasonable.
- The weak initialization has ramifications well beyond early boot; consider services which must start early and _continue running_ - init systems, DBus, FUSE filesystems, etc. These then become more vulnerable to hash-DOS over the entire runtime of the system, and are crucial system components.
- There's no way for the hash initialization, if initially weak, to become strong via reseeding later.

In essence, I feel that this change weakens `HashMap` for the explicit purpose of supporting an antipattern with detrimental effects on the ecosystem, when the proper way of handling the issue would be to recommend using `BTreeMap` if entropy may not be available.

**Comment by nejucomo** [comment]
Chiming in to heartily concur with @eternaleye 's post. Also imagine you are a new-to-rust developer and you read this patch's ``HashMap`` documentation:

> This means that the ordering of the keys is randomized, but makes the tables more resistant to denial-of-service attacks (Hash DoS). No guarantees are made to the quality of the random data. The implementation uses the best available random data from your platform at the time of creation. This behavior can be overridden with one of the constructors.

Now, I either need to be concerned about DoS for my use case, or I do not. If I do care, how can I evaluate whether or not ``HashMap`` helps my use case?

If the documentation were extended to spell out the conditions in which the entropy (and thus DoS) protection is weak, can I determine whether or not my use case is in that condition?  (A panic is one clear way to determine if that's the case, and for the use cases that require ``HashMap`` DoS protection, this provides fail-closed security.)

If it's not possible to evaluate or determine when or how this DoS protection comes into effect, then isn't its value largely diminished? That's akin to ignoring if your HTTPS connection has a valid cert or not because "encryption may help prevent network eavesdroppers".

Meanwhile, for the other use case of early startup, if DoS protection does not matter, then why use a library that sometimes provides that feature anyway, instead of a simpler alternative?

**Comment by alexcrichton** [comment]
@eternaleye this was why [we concluded](https://github.com/rust-lang/rust/pull/33086#issuecomment-217043251) that librand should be documented as not being suitable outside the standard library.

**Comment by eternaleye** [comment]
@alexcrichton: But that doesn't address my point: This change directly _encourages_ using `HashMap` in early-boot environements, which (for code which _continues running_ after early-boot) means it retains a vulnerability to hash-DOS. Using `librand` outside `std` has nothing to do with it; using `HashMap` in early boot is the problem.

**Comment by cardoe** [comment]
@eternaleye Many things rely on `HashMap` internally. How do you propose that people handle this situation to avoid that hidden dependency on `HashMap`? Additionally how do you propose that users are inherently aware of this and can avoid it? You mention init systems in your post so I'll link you to systemd, https://github.com/systemd/systemd/commit/97768fc5746ea97cb2d3ac6854b4fc7ce1c14f24 which has had the same behavior since 2014. All the other daemons you mentioned would come up after the random pool is initialized in most use cases and render your arguments invalid.

@nejucomo Hash tables that accept unbound data sets (where a user can put any data in) are vulnerable to DoS attacks if they are not seeded.

While the original bug report asked for some crates to move away from using `HashMap` and then later on asked that an application can decide how `HashMap`'s are seeded this was the solution that was accepted.

@eternaleye 
> But that doesn't address my point: This change directly encourages using HashMap in early-boot environments, which (for code which continues running after early-boot) means it retains a vulnerability to hash-DOS. Using librand outside std has nothing to do with it; using HashMap with insufficient entropy is the problem.

You assume that every daemon is accepted unbounded data that is started before the random pool is initialized. You also assume that existing crate authors are aware of this and have the ability to avoid using `HashMap`. There is almost no ability with the current Rust crate environment to write a program past "Hello World" that can be used without a `HashMap` wanting to be initialized. I would hardly call this "directly encouraging using HashMap".

Honestly I would love to see the people opposed to this change propose a solution that's acceptable to the Rust team because I would be willing to implement it. Just like I was willing to implement the solution that was requested of me initially.

**Comment by eternaleye** [comment]
@cardoe: Many things use `/proc`, or `/sys`, or other things that may not be available at early-boot internally. Call `fexecve(3)` on a kernel without `execveat(2)` (anything prior to 3.19) without /proc mounted? Boom. Kernel has `execveat(2)` but glibc is too old to use it? Boom. Early boot is a constrained environment that requires significant care; papering over "`HashMap` may block" by changing it to "`HashMap` may be vulnerable to DOS" doesn't fix the issue. Instead, it allows the issue's ramifications to propagate _beyond early boot_.

And yes, systemd does something I consider incorrect there, especially as it _does_ allow unbounded data: `systemctl set-environment` among other things. (like the DBus interface for defining transient units...)

**Comment by cardoe** [comment]
@eternaleye Enough grandstanding and throwing out irrelevant examples. Propose an actual solution to the issue at hand and we can move forward. I am more than willing to implement another solution.

**Comment by eternaleye** [comment]
@cardoe: I am not grandstanding, and I'm rather insulted by the implication. In addition, the examples are not irrelevant. Fundamentally, early but is not a POSIX environment, and it makes no guarantees whatsoever of what is available. `std` relies on more than early boot provides.

`/dev/urandom` may not exist, as the device node may not have been created. Various C library calls may not work, because `/proc` and `/sys` are not mounted. The list goes on.

In addition, while blocking when unable to get randomness makes `HashMap` itself unusable in early boot, accepting sub-par randomness allows using `HashMap` in early boot to silently reduce the security of the system later in its runtime, after it's fully initialized.

Fundamentally, early boot is a `#![no_std]` environment - some crates like `alloc` are usable, but others are not. This "solution" _isn't_ - it's an ad-hoc patch job for a very small part of a larger issue, with real and problematic side effects.

**Comment by briansmith** [comment]
> Propose an actual solution to the issue at hand and we can move forward. I am more than willing to implement another solution.

* By default OsRng should just use `getrandom` and not fall back to reading from /dev/urandom unless `getrandom` returns `ENOSYS`. i.e. by default, the old blocking behavior is restored.
* Much like the way the application can configure `panic!` to abort on failure, a similar mechanism should be implemented where the application (and in particular not a library) can say that `OsRng` should never block. This mechanism would be intended only for use by specialized apps that are designed to run before the OS ensures that `getrandom()` would never block.

Note that such design follows the "secure by default" principle.

In particular, consider an application that invokes the `getrandom` syscall directly and, if the system doesn't return `ENOSYS`, filters out the `open[at]`, `read`, etc. syscalls using seccomp-bpf. Such an application would work correctly with the previous behavior but not with the currently-commited behavior. (Thus, the already-committed change in this PR is a backward compatibility regression.) With the strategy I suggested above, the backward compatibility would be restored.

**Comment by cardoe** [comment]
@eternaleye 
> Fundamentally, early boot is not a POSIX environment, and it makes no guarantees whatsoever of what is available. std relies on more than early boot provides.

Where in the world do you get this crazy notion from? I'd love to see some document to back this assertion up.

> /dev/urandom may not exist, as the device node may not have been created. Various C library calls may not work, because /proc and /sys are not mounted. The list goes on.

So you're arguing that Rust can not be used as an init system then? So Rust is acceptable to bare metal embedded systems and to fully stood up POSIX environments only.

> Fundamentally, early boot is a #![no_std] environment - some crates like alloc are usable, but others are not. This "solution" isn't - it's an ad-hoc patch job for a very small part of a larger issue, with real and problematic side effects.

That's absolutely absurd assertion to make. #![no_std] is clearly not for that use case to imply it is absolutely false. I encourage you to read the Rust book again. https://doc.rust-lang.org/book/no-stdlib.html

The Rust developers pointed me in this direction because users have the ability to use a different hash function with `HashMap` should they need/want to. But users of basic crates that provide things like argument parsing for example have no way to specify the hash function used for the `HashMap` implementation. The point here is you can opt-in to what you need.

@briansmith 
Saying "revert the change" does not count as proposing a solution. Plain and simple if you need to spew unbound data at `HashMap` and you are concerned that your application is started before the random pool is initialized use the existing mechanisms to supply your own hasher function. I will gladly update the documentation to reflect that.

**Comment by briansmith** [comment]
> @briansmith 
> Saying "revert the change" does not count as proposing a solution. Plain and simple if you need to spew unbound data at HashMap and you are concerned that your application is started before the random pool is initialized use the existing mechanisms to supply your own hasher function. I will gladly update the documentation to reflect that.

The first bullet of what I suggested was indeed effectively "revert the change." Such reversion is needed to undo the compatibility regression anyway. However, the second bullet of what I suggested was the necessary step to get the behavior you want: an opt-in mechanism that any application can make to get the behavior you want.

**Comment by cardoe** [comment]
As I've said on the previous issues. Provide another hasher function in the standard library and steer crate authors to the right implementation for the right use case. I get the "secure by default" attitude but people would similarly find it over the top to establish a full TLS connection and teardown for 2 bytes of data. Its all about the right tool for the right job. In most of the crates I see using `HashMap`, using a fully seeded SipHash is over the top.

**Comment by cardoe** [comment]
@briansmith 
> The first bullet of what I suggested was indeed effectively "revert the change." Such reversion is needed to undo the compatibility regression anyway. However, the second bullet of what I suggested was the necessary step to get the behavior you want: an opt-in mechanism that any application can make to get the behavior you want.

As was pointed out by @alexcrichton that was discussed and the team decided to proceed forward.

**Comment by eternaleye** [comment]
@cardoe: You are putting words in my mouth. Please stop that.

> Where in the world do you get this crazy notion from?

...the fact that various things required by POSIX can be absent during early boot?
- Mandatory utilities such as `basename(1)` may not be present (consider initramfs)
- `/dev` may not contain `/dev/console` or `/dev/tty`, which [POSIX mandates](http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap10.html)

> So you're arguing that Rust can not be used as an init system then? So Rust is acceptable to bare metal embedded systems and to fully stood up POSIX environments only.

No, I said what I meant and I meant what I said. Some crates are valid and some are not. `std` assumes all of its component crates are valid; `#![no_std]` allows more fine-grained handling. The very link you gave bears this up, starting with the very first paragraph:

> Rust’s standard library [...] assumes support for various features of its host system [...] There are systems that do not have these features, however, and Rust can work with those too! To do so, we tell Rust that we don’t want to use the standard library via an attribute: #![no_std].

One can then load the individual crates that are usually bundled behind the `libstd` facade: `liballoc` (which provides `Box<T>`, `Rc<T>`, and more), `libcontainers` (which provides, among other things, `BTreeMap`, `HashMap`, etc), and so forth.

`#![no_std]` is in no way only "acceptable to bare metal embedded systems" - that's flat out incorrect.

It's not black and white "`std` or nothing" - it's "open up the `std` black box and only use what's needed."

Unless the _full_ set of assumptions behind `std` are upheld, the facade is not the correct abstraction to use - but the individual crates behind the facade have simpler assumptions, and can be used even if the full power of `std` is off the table.

**Comment by cardoe** [comment]
@eternaleye 
> ...the fact that various things required by POSIX can be absent during early boot?

> Mandatory utilities such as basename(1) may not be present (consider initramfs)
/dev may not contain /dev/console or /dev/tty, which POSIX mandates

Where in libstd does it call out to the basename binary or open /dev/console? We're talking about a compiled program and a shell environment. You're still discussing apples to oranges.

**Comment by eternaleye** [comment]
I never said it did either. I said early boot is not a POSIX environment, and gave two examples I was able to find quickly, rather than spend hours wading through the spec for more targeted ones.

**Comment by tbu-** [comment]
@cardoe I think that you're overly aggressive about this.

**Comment by stouset** [comment]
Sorry if I'm late to the party — I came here after reading the patch notes for 1.10 and being concerned about the notion about reverting to `/dev/urandom` if `getrandom(2)` blocks.

It seems to me this is a poor security tradeoff. Rust programs that are intended for early-boot environments are surely a somewhat-rare use case. In these cases, such binaries should be forced to opt-in to reduced security — in some cases it's probably not a big deal, but for others (long-lived programs with unbounded HashMaps) it can and will cause security issues.

This isn't a small project that will barely see the light of day — as Rust maintainers, you're making security decisions and tradeoffs that affect *all* of your users. And trading away strong security by default to favor rare use cases (early-boot programs that are short-lived and/or have bounded datasets in HashMaps) does a disservice to the majority.

**Comment by brson** [comment]
@stouset I suggest filing a bug about it. Comments here will be ineffective.

I'm also a bit worried about this, but it's important to remember that the only thing this affects is the order things are inserted into hash tables. That is, the only attack this could enable is hash table DDOS. It's not a weakening of random numbers in Rust in general.

**Comment by stouset** [comment]
Yes, I realize that's the case. This is, to be clear, a somewhat rare use-case that most Rust developers will likely never encounter.

**Pull request merged**